### PR TITLE
feat: Support auto power control

### PR DESCRIPTION
**Pull request by huyongqii**
## Summary by CodeRabbit

- **New Features**
  - Introduced comprehensive power control plugin with prediction-based scaling, IPMI/SSH node management, and a Python predictor service.
  - Added hooks and RPCs for updating node power states and registering network interfaces.
  - Enabled job-level energy and resource monitoring with detailed metrics and synchronous InfluxDB writes.
  - Added configuration options for power control, job monitoring, and network interface management.

- **Enhancements**
  - Refined energy monitoring focus from tasks to jobs for improved metrics aggregation.
  - Improved plugin interfaces, protocol definitions, and configuration validation to support new power states and job metrics.
  - Simplified metric collection methods to return default values on error, enhancing stability.
  - Enhanced node state display to include power state information.

- **Bug Fixes**
  - Improved error handling and logging across plugins and monitoring components.

- **Refactor**
  - Replaced task-based monitoring with job-based monitoring in energy plugin.
  - Removed buffering and batch writes in database operations, adopting direct synchronous writes.
  - Streamlined metric retrieval methods by eliminating error returns and logging internally.

- **Documentation**
  - Updated configuration files to rename fields, remove deprecated options, and reflect new features.

- **Chores**
  - Added new Python and Go files implementing prediction models, power management logic, and IPMI-based control.
  - Removed obsolete task monitoring code and associated structures.

### CODE REVIEW

File: internal/ccontrol/CmdArgParser.go [hunk 1]
```diff
@@ -235,6 +237,18 @@ var (
 			}
 		},
 	}
+	powerCmd = &cobra.Command{
+		Use:   "power [flags] craned_id[,craned_id...]",
+		Short: "Control node power state (wake/sleep/on/off)",
+		Long:  "Execute power control actions on specified craneds",
+		Args:  cobra.ExactArgs(1),
+		Run: func(cmd *cobra.Command, args []string) {
+			cranedIds := strings.Split(args[0], ",")
```
[the remark this answers is not in the document]
Comment (huyongqii) [reply internal/ccontrol/CmdArgParser.go:246]: 指的是这个意思对么`cranedIds, err := util.ParseStringParamList(args[0], ",")`

File: plugin/powerControl/plugin.go [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+	go func() {
+		scanner := bufio.NewScanner(stdout)
+		for scanner.Scan() {
+			log.Info("Predictor: ", scanner.Text())
+		}
+	}()
+
+	go func() {
+		scanner := bufio.NewScanner(stderr)
+		for scanner.Scan() {
```
Comment (L-Xiafeng) [inline plugin/powerControl/plugin.go:77]: Predictor应该自己负责输出？

File: internal/cplugind/server.go [hunk 3]
```diff
@@ -133,3 +133,63 @@ func (pd *PluginDaemon) NodeEventHook(ctx context.Context, req *protos.NodeEvent
...
+	c.Start()
+
+	if c.Get("ok").(bool) {
+		reply.Ok = true
+	} else {
+		reply.Ok = false
+		reply.Error = c.Get("error").(string)
+	}
+
+	return reply, nil
```
Comment (Nativu5) [inline internal/cplugind/server.go:194]: 如果是同步 Hook，最好在名字上做一些区分。

**Review by Nativu5 — CHANGES_REQUESTED** [review]

File: plugin/energy/energy.go [hunk 4]
```diff
@@ -132,8 +132,16 @@ func (p EnergyPlugin) CreateCgroupHook(ctx *api.PluginContext) {
...
+	requestCpu := req.Resource.AllocatableResInNode.CpuCoreLimit
+	requestMemory := req.Resource.AllocatableResInNode.MemoryLimitBytes
+	boundGPUs := getBoundGPUs(req.Resource.DedicatedResInNode, globalMonitor.config.Monitor.GPUType)
+	resourceRequest := monitor.ResourceRequest{
+		ReqCPU:    requestCpu,
+		ReqMemory: requestMemory,
+		ReqGPUs:   boundGPUs,
+	}
+
+	globalMonitor.monitor.TaskMonitor.Start(req.TaskId, req.Cgroup, resourceRequest)
```
Comment (Nativu5) [inline plugin/energy/energy.go:144]: 后续 Crane 要使用 Job 作为任务的翻译了，新代码可以把相关变量命名同步到 Job

**Review by Nativu5 — CHANGES_REQUESTED** [review]

File: internal/cplugind/server.go [hunk 5]
```diff
@@ -133,3 +133,63 @@ func (pd *PluginDaemon) NodeEventHook(ctx context.Context, req *protos.NodeEvent
...
+
+	c := api.NewContext(ctx, req, api.GetCranedListHook, &hs)
+	c.Start()
+
+	if ids, ok := c.Get("craned_ids").([]string); ok {
+		reply.CranedIds = ids
+	}
+
+	return reply, nil
+}
```
[the remark this answers is not in the document]
Comment (huyongqii) [reply internal/cplugind/server.go:174]: 给调度器查询当前可用的节点（开机且未睡眠的节点），这块后续需要跟奕聪聊下，因为需要那边接入这个函数
目前有个问题是：每次调度都需要获得当前可用的节点，这个必须要是同步的，在我们目前的调度流程中，这个耗时是否可以接受

File: internal/cplugind/server.go (hunk 3 above)
Comment (huyongqii) [reply internal/cplugind/server.go:194]: 给调度器查询当前可用的节点（开机且未睡眠的节点），这块后续需要跟奕聪聊下，因为需要那边接入这个函数
目前有个问题是：每次调度都需要获得当前可用的节点，这个必须要是同步的，在我们目前的调度流程中，这个耗时是否可以接受

File: plugin/powerControl/plugin.go (hunk 2 above)
Comment (huyongqii) [reply plugin/powerControl/plugin.go:77]: 我后续改下，让Predictor单独输出到一个日志文件中

**Comment by Nativu5** [comment]
现在是否还是必须依赖同步 hook 呢？如果需要的话可以给 hook 换个名称，比如 SyncHook 或者改成 Call 之类的词。

File: plugin/energy/pkg/db/influxdb.go [hunk 6]
```diff
@@ -21,28 +19,42 @@ type InfluxDB struct {
...
+	jobBucket  string
+	enabled    *config.Enabled
 }
 
 func NewInfluxDB(config *config.Config) (*InfluxDB, error) {
-	client := influxdb2.NewClient(config.DB.InfluxDB.URL, config.DB.InfluxDB.Token)
-	if _, err := client.Ping(context.Background()); err != nil {
-		return nil, fmt.Errorf("failed to ping InfluxDB: %v", err)
+	maxRetries := 3
+	retryInterval := 5 * time.Second
```
Comment (Nativu5) [inline plugin/energy/pkg/db/influxdb.go:28]: 这个会不会作为非导出的const比较好？

File: plugin/energy/pkg/monitor/job.go [hunk 7]
```diff
@@ -287,3 +285,42 @@ func (t *Task) logTaskStats() {
...
+		totalMemoryRequest += float64(job.resourceRequest.ReqMemory)
+		totalRuntime += time.Since(job.data.StartTime).Minutes()
+	}
+	m.mutex.RUnlock()
+
+	if metrics.JobCount > 0 {
+		metrics.ReqCPURate = totalCPURequest / cpuCores
+		metrics.AvgReqCPUPerJob = totalCPURequest / float64(metrics.JobCount)
+		metrics.ReqMemoryRate = totalMemoryRequest / totalMemoryBytes
+		metrics.AvgReqMemoryGBPerJob = (totalMemoryRequest / float64(metrics.JobCount)) / (1024 * 1024 * 1024) // 转换为GB
```
Comment (Nativu5) [inline plugin/energy/pkg/monitor/job.go:314]: 注意中文注释

File: plugin/energy/pkg/monitor/job.go [hunk 8]
```diff
@@ -287,3 +285,42 @@ func (t *Task) logTaskStats() {
...
+		metrics.AvgJobRuntime = totalRuntime / float64(metrics.JobCount)
+	}
+
+	log.Infof("Job metrics: %+v", metrics)
+	log.Infof("ReqCPURate: %f", metrics.ReqCPURate)
+	log.Infof("AvgReqCPUPerJob: %f", metrics.AvgReqCPUPerJob)
+	log.Infof("ReqMemoryRate: %f", metrics.ReqMemoryRate)
+	log.Infof("AvgReqMemoryGBPerJob: %f", metrics.AvgReqMemoryGBPerJob)
+	log.Infof("AvgJobRuntime: %f", metrics.AvgJobRuntime)
```
Comment (Nativu5) [inline plugin/energy/pkg/monitor/job.go:323]: 可能日志级别要降低一下，这样可能日志输出太多了。

**Review by Nativu5 — CHANGES_REQUESTED** [review]

File: plugin/energy/pkg/db/influxdb.go (hunk 6 above)
Comment (huyongqii) [reply plugin/energy/pkg/db/influxdb.go:28]: OK

File: plugin/energy/pkg/db/influxdb.go [hunk 9]
```diff
@@ -51,181 +63,141 @@ func NewInfluxDB(config *config.Config) (*InfluxDB, error) {
...
-			}
-		} else {
-			db.bufferMu.Unlock()
-		}
-	}
-}
-
-func (db *InfluxDB) SaveTaskEnergy(taskData *types.TaskData) error {
-	log.Infof("Saving task energy data for task: %d, node: %s", taskData.TaskID, taskData.NodeID)
+func (db *InfluxDB) SaveJobEnergy(jobData *types.JobData) error {
```
[the remark this answers is not in the document]
Comment (huyongqii) [reply plugin/energy/pkg/db/influxdb.go:166]: 回顾了下这块，目前job metrcis这块也是按1min采样一次，与node metrics不同的是，每次采样不会直接写入influxdb，也不会一直保存，下一次采样会聚合计算上一次采样，所以metrics占用的内存不会随着job的运行时间而增加，因为没有保存job在运行过程中历史信息，保存是最终聚合计算后的一些变量，如消耗的能耗，平均cpu、内存利用率这些，所以占用内存很小，同步写不会影响性能

**Comment by huyongqii** [comment]
> 现在是否还是必须依赖同步 hook 呢？如果需要的话可以给 hook 换个名称，比如 SyncHook 或者改成 Call 之类的词。

目前只有一个同步hook:`GetCranedByPowerStateHookSync` 暂时还用不到

File: plugin/powerControl/action.go [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+
+			if end < len(allowedNodes) {
+				log.Infof("Waiting %d seconds before processing next power off batch...",
+					c.config.IPMI.PowerOffBatchIntervalSeconds)
+				time.Sleep(powerOffBatchInterval)
+			}
+		}
+	}()
+
+	return nil
```
[the remark this answers is not in the document]
Comment (Nativu5) [reply plugin/powerControl/action.go:93]: 这个没有 wait 是对的吗？

File: plugin/powerControl/predictor/model.py [hunk 11]
```diff
@@ -0,0 +1,166 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import math
+
```
Comment (L-Xiafeng) [inline plugin/powerControl/predictor/model.py:5]: 模型不应该放在我们仓库里吧？

File: plugin/powerControl/action.go [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+	allowedNodes := c.filterExcludedNodes(nodeIDs)
+	for _, nodeID := range allowedNodes {
+		if err := c.sleepNode(nodeID); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *PowerManager) powerOnNodes(nodeIDs []string) error {
```
Comment (L-Xiafeng) [inline plugin/powerControl/action.go:42]: poweron没有batch限制吗

Comment (huyongqii) [reply plugin/powerControl/action.go:42]: 是的，需求中没对开机进行限制

File: plugin/powerControl/predictor/model.py (hunk 11 above)
Comment (huyongqii) [reply plugin/powerControl/predictor/model.py:5]: 目前预测服务是跟随PowerControl一块启动的，所以需要model也放在这里

Comment (L-Xiafeng) [reply plugin/powerControl/predictor/model.py:5]: 这两个东西应该分开，不能绑定在一起，预测这个应该是可选的

File: plugin/powerControl/action.go (hunk 12 above)
Comment (L-Xiafeng) [reply plugin/powerControl/action.go:42]: 这个应该也是有的，可以再确认下

Comment (huyongqii) [reply plugin/powerControl/action.go:42]: 没有

Comment (huyongqii) [reply plugin/powerControl/action.go:42]: 跟马老师反应了下这个问题，还是需要实现下

Comment (huyongqii) [reply plugin/powerControl/action.go]: 已完善

**Review by Nativu5 — APPROVED** [review]
